text/libobject-refactor.md

```diff
@@ -0,0 +1,197 @@
+with the kernel play. Indeed, some objects present in a `.vo` can be
+understood as "logical" objects, as they will be registered with the
+kernel.
+
+A large set of objects are meta-logical: they are not registered with
+the kernel, but still essential to Coq operation. Such objects are
+notations, instance declarations, hints, etc...
+
+Indeed it is a common source of discussion where the data for logical
+objects should be stored. Some data will be stored invariant in the
+kernel, but some other data, like documentation or location, has to
+live outside it.
+
+# Part I: common meta-data
+
+For this part, we propose to decorate the `Libobject.t` type with a
+`Data.t` structure. This approach is modelled after the approach used
+in `Declare` for constant information, which has proven successful in
+reducing code duplication, uniformizing the API, and making it more
+extensible (see for example per-definition typing flags).
+
+In this preliminary proposal, `Data.t`, will contain:
+
+- `loc`: Location of the object, if it is coming from a source file
+- `doc`: String containing the documentation of the object, to be extended
+- `name`: Name of the object, as an `Id.t`, if not anonymous
+- `category`: Already exiting in `Libobject`, used for import, could be also used by `Search`.
+
+A draft implementation is available in PR (number to be added ASAP)
+
+This changes affect the lifetime in the following way:
+
+## Creation
+
+Clients must provide the metadata when registering the objects.
+
+## Loading
+
+`Declaremods` already stores the objects in a map indexed by their
+prefix, we propose to:
+
+- create a reverse map or maps to allow clients to query the `Data.t`
+  efficiently. In particular we could have reverse maps for
+  categories, names, locations. For now we implement a reverse map for
+  names, but more options could be desirable after more experimentation.
+
+- expose the regular map for anonymous objects so clients can explore.
+
+## Extra
+
+It would be desirable to allow clients to register as to allow
+incremental indexing of objects, of for example, to maintain a
+location-based table.
+
+Another interesting topic is versioning of objects. While preliminary,
+we could include versioning information in the objects as for example
+to allow to have hybrid Coq 8/9 objects, or several versions of the
+same `Definition` for example.
+
+# Part II: behavior specification
+
+The second part of this CEP, not yet implemented, is to phase out the
+`declare_object` API in favor a set of `Behavior.t` traits that an
+object can implement, in a functional way.
+
+The main goal here is to allow the upper layer to control the state,
+and have more flexibility in updating the state. This has many
+applications.
+
+In particular, we can distinguish a few common behaviors:
+
+- `Kernel` : objects with this trait provide a function to register with a `safe_env`
+- `Named` : objects with this trait can register with the `Nametab`, in a uniform way
```

Comment:
My understanding is that currently nametab itself is not uniform (there are separate tables for globref, modules, ltacs, some other stuff) so this may require a nametab redesign first.

Reply:
FTR I think that unifying the various nametabs is gong the wrong way. It's a feature that each plugin can define its own set of names, so the design critically needs to account for this.

Reply:
Yes, the idea is that when an object declares a nametab trait, it does so with their own nametab. Refactoring is indeed needed, you can see current progress here, but need to advance more https://github.com/ejgallego/coq/blob/f059e5a57e95c5c80a3c9026e91ba91936d90db5/library/nametab.mli

Ideally, if we think this is a good idea we would do such a change in two steps:
- we would first define the trait, which takes a module as a parameter and performs the imperative state
- we would then push the trait to be functional, and have `Declaremods` own the nametab.

Reply:
Clarified a bit.